src/reconstruction.jl

```diff
@@ -0,0 +1,214 @@
+	for φ=1:Φ
+		r.mini[φ] <= idx[φ] <= r.maxi[φ] || return false
+ 	end
+ 	return true
+end
+Base.CartesianIndices(r::Region{Φ}) where Φ =
+	 CartesianIndices{Φ,NTuple{Φ,UnitRange{Int}}}(([r.mini[φ]:r.maxi[φ] for φ=1:Φ]...,))
+
+
+function inner_region(βs::Vector{CartesianIndex{Φ}}, fsize) where Φ
+	mini = Int[]
+	maxi = Int[]
+	for φ = 1:Φ
+		js = map(β -> β[φ], βs) # jth entries
+		mi,ma = extrema(js)
+		push!(mini, 1 - min(mi, 0))
+		push!(maxi,fsize[φ] - max(ma, 0))
+	end
+	return Region{Φ}((mini...,), (maxi...,))
+end
+
+function project_inside(α::CartesianIndex{Φ}, r::Region{Φ}) where Φ
+	CartesianIndex(mod.(α.I .-1, r.maxi).+1)
+end
+
+"""
+	 SpatioTemporalEmbedding{T,Φ,BC,X} <: AbstractSpatialEmbedding{T,Φ,BC,X} → `embedding`
+A spatio temporal delay coordinates structure to be used as a functor.
+
+	embedding(rvec, s, t, α)
+Operates inplace on `rvec` and reconstructs vector from spatial timeseries `s` at
+timestep `t` and cartesian index `α`.
+
+## Constructors
+The structure can be created directly by calling
+
+	SpatioTemporalEmbedding{T,Φ,BC,X}(τ,β,fsize)
+where `T` is the `eltype` of the timeseries, `Φ` the spatial dimension of the system,
+`BC` the boundary condition type and `X` the length of reconstructed vectors.
+Arguments `τ` and `β` are Vectors of `Int` and `CartesianIndex` that contain
+all points to be included in the reconstruction in *relative* coordinates
+and `fsize` is the size of each state in the timeseries.
+
+Note that this is a forward embedding and `τ` needs to be sorted in ascending order.
+The most recent values in the reconstruction are located at the end of each vector.
+
+A simpler constructor for convenience
+
+	SpatioTemporalEmbedding(s, D, τ, B, k, ::Type{<:AbstractBoundaryCondition})
+Takes as arguments the spatial timeseries `s` and reconstructs
+`B` spatial shells separated by `k` points around each point
+and repeats this for `D` past timesteps separated by `τ` each.
+"""
+struct SpatioTemporalEmbedding{T,Φ,BC,X} <: AbstractSpatialEmbedding{T,Φ,BC,X}
+  	τ::Vector{Int}
+	β::Vector{CartesianIndex{Φ}}
+	inner::Region{Φ}  #inner field far from boundary
+	whole::Region{Φ}
+    boundary::BC
+
+	function SpatioTemporalEmbedding{T,Φ,BC,X}(τ,β,bc,fsize) where {T,Φ,BC,X}
+		inner = inner_region(β, fsize)
+		whole = Region((ones(Int,Φ)...,), fsize)
+		return new{T,Φ,BC,X}(τ,β,inner,whole, bc)
+	end
+end
+const STE = SpatioTemporalEmbedding
+
+function SpatioTemporalEmbedding(
+		s::AbstractArray{<:AbstractArray{T,Φ}},
+		D, τ, B, k, boundary::BC
+		) where {T,Φ, BC<:AbstractBoundaryCondition}
+	@assert issorted(τ) "Delays need to be sorted in ascending order"
+    #"ConstantBoundary condition value C needs to be the same type as values in s"
+    @assert BC <: PeriodicBoundary || typeof(boundary.C) == T "typeof(boundary.C) == eltype(s[1])"
+	X = (D+1)*(2B+1)^Φ
+	τs = Vector{Int}(undef,X)
+	βs = Vector{CartesianIndex{Φ}}(undef,X)
+	n = 1
+	for d=0:D, α = Iterators.product([-B*k:k:B*k for φ=1:Φ]...)
```

Comment:
`Iterators.product([-B*k:k:B*k for φ=1:Φ]...)` can't be right?

First of all you are iterating over a single vector. What does the `product` offer you, which is useful over many iterators.
Secondly, the expression `thing(a) for a in otherthing` is an iterator by itselff. Youy don't have to collect it in a vector `[]` and then splice it with `...`.

Reply:
Disregard this comment. I know now why this is done.